### DISCUSSION TITLE
No per-exchange publishing rates for MQTT-originated messages in RabbitMQ Management

### DISCUSSION THREAD

**Discussion by johanrhodin** [rabbitmq/rabbitmq-server#8515] [Other] [answered]
After updating from 3.11.17 to 3.12 the message rates for MQTT-based publishers are gone. It will now look like exchanges are idle  when they are not, and that messages just appears in the queues without related publishes.

<img width="626" alt="Screen Shot 2023-06-09 at 16 34 45" src="https://github.com/rabbitmq/rabbitmq-server/assets/460369/8ec36c3d-88b5-44b2-a99c-26eef50ead0e">
<img width="1245" alt="Screen Shot 2023-06-09 at 16 34 31" src="https://github.com/rabbitmq/rabbitmq-server/assets/460369/5097761a-631c-4210-9474-96d9c36247fc">

Is this as designed? If so it should be mentioned in the release notes or at least the accompanying blog post.

**Comment by nsurleraux-railnova** [comment]
@johanrhodin Did you solve that? Maybe this should be converted into an issue as it may be blocking to upgrade a RabbitMQ cluster.

**Reply by michaelklishin** [reply]
It's unlikely that this metric will be coming back before at least one more protocol implementation (most likely AMQP 1.0) will switch to the same "chanenl-less" implementation and the metrics subsystem acommodates both of their needs.

With Prometheus, you can use certain connection metrics and aggregate across them in Grafana to have a sense of the total rate or data flow. Any cluster where multiple virtual hosts are used would have to do this even with versions earlier than 3.12.

**Comment by michaelklishin** [comment]
This is expected. All management UI metrics that are tied to channels are no longer emitted by the MQTT implementation. Per-exchange ingress rate is one of those metrics.

There are other metrics, such as per-connection rates, that are still available and should be exposed via the Prometheus scraping endpoint.

**Reply by ansd** [reply]
Yes, see also https://groups.google.com/g/rabbitmq-users/c/XVR1OGkT0Bo/m/slWmsVFhCAAJ

**Comment by michaelklishin** [comment, marked as answer]
Release notes now feature the following under the `Compatibility` section:

``` markdown
### MQTT Client Metrics

Some metrics displayed by the management UI are no longer available for MQTT clients.
They are computed for AMQP 0-9-1 channels, which MQTT plugin no longer uses. They may be
reintroduced in future releases when more protocol implementations (e.g. AMQP 1.0) are
refactored the same way the MQTT plugin was for this release.

Connection-level metrics such as ingress and egress data flows are still available,
and so are the metrics of queues used by MQTT subscribers.
```

**Reply by johanrhodin** [reply]
OK! I see it here: https://github.com/rabbitmq/rabbitmq-server/commit/db92a19aab6e4abedf603cfb32200480a9e35ebe 🙏
